File: scm-java/src/main/scheme/SchemeThread.java
package scheme;

public class SchemeThread {
    public static final ThreadLocal<SchemeThread> currentThread = new ThreadLocal<>();

    public enum State { NEW, STARTED, TERMINATED }

    public Thread thread;
    public Object result = Value.NIL;
    public Lambda lambda;
    public Modules modules;
    public Object name = Value.F;
    public Object specific = Value.NIL;
    public Object exception;
    public SchemeError originalError;
    public volatile boolean terminated;
    public State state = State.NEW;

    public SchemeThread(Lambda lambda, Modules modules) {
        this.lambda = lambda;
        this.modules = modules;
    }

    @Override
    public String toString() {
        return name != Value.F ? "#<thread " + Value.displayRep(name) + ">" : "#<thread>";
    }
}

File: scm-csharp/scheme/SchemeThread.cs
using System.Threading;
using System.Threading.Tasks;

namespace scheme;

public enum SchemeThreadState { NEW, STARTED, TERMINATED }

public class SchemeThread
{
    [ThreadStatic]
    public static SchemeThread? CurrentThread;

    public Task? task;
    public object result = Value.NIL;
    public Lambda? lambda;
    public Modules modules;
    public object name = Value.F;
    public object specific = Value.NIL;
    public object? exception;
    public SchemeError? originalError;
    public volatile bool terminated;
    public SchemeThreadState state = SchemeThreadState.NEW;

    public SchemeThread(Lambda? lambda, Modules modules)
    {
        this.lambda = lambda;
        this.modules = modules;
    }

    public override string ToString() =>
        !name.Equals(Value.F) ? "#<thread " + Value.DisplayRep(name) + ">" : "#<thread>";
}
